tracee-ebpf/tracee/internal/bufferdecoder/decoder.go

```diff
@@ -0,0 +1,301 @@
+// sequences and the user-defined structs.
+//
+// The package favors efficiency over flexibility. The provided API
+// allows fast decoding of byte sequence sent by the Tracee eBPF program from
+// kernel-space to user-space.
+package bufferdecoder
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/aquasecurity/tracee/pkg/external"
+)
+
+type EbpfDecoder struct {
+	buffer []byte
+	cursor uint32
+}
+
+// New creates and initializes a new EbpfDecoder using rawBuffer as its initial content.
+// The EbpfDecoder takes ownership of rawBuffer, and the caller should not use rawBuffer after this call.
+// New is intended to prepare a buffer to read existing data from it, translating it to protocol defined structs.
+// The protocol is specific between the Trace eBPF program and the Tracee-eBPF user space application.
+func New(rawBuffer []byte) *EbpfDecoder {
+	return &EbpfDecoder{
+		buffer: rawBuffer,
+		cursor: 0,
+	}
+}
+
+// BuffLen returns the total length of the buffer owned by decoder.
+func (decoder *EbpfDecoder) BuffLen() int {
+	return len(decoder.buffer)
+}
+
+// ReadAmountBytes returns the total amount of bytes that decoder has read from its buffer up until now.
+func (decoder *EbpfDecoder) ReadAmountBytes() uint32 {
+	return decoder.cursor
+}
+
+// DecodeContext translates data from the decoder buffer, starting from the decoder cursor, to protocol.Context struct.
+func (decoder *EbpfDecoder) DecodeContext(ctx *Context) error {
+	offset := decoder.cursor
+	if len(decoder.buffer[offset:]) < 102 {
+		return fmt.Errorf("can't read context from buffer: buffer too short")
+	}
+	ctx.Ts = binary.LittleEndian.Uint64(decoder.buffer[offset : offset+8])
+	ctx.CgroupID = binary.LittleEndian.Uint64(decoder.buffer[offset+8 : offset+16])
+	ctx.Pid = binary.LittleEndian.Uint32(decoder.buffer[offset+16 : offset+20])
+	ctx.Tid = binary.LittleEndian.Uint32(decoder.buffer[offset+20 : offset+24])
+	ctx.Ppid = binary.LittleEndian.Uint32(decoder.buffer[offset+24 : offset+28])
+	ctx.HostPid = binary.LittleEndian.Uint32(decoder.buffer[offset+28 : offset+32])
+	ctx.HostTid = binary.LittleEndian.Uint32(decoder.buffer[offset+32 : offset+36])
+	ctx.HostPpid = binary.LittleEndian.Uint32(decoder.buffer[offset+36 : offset+40])
+	ctx.Uid = binary.LittleEndian.Uint32(decoder.buffer[offset+40 : offset+44])
+	ctx.MntID = binary.LittleEndian.Uint32(decoder.buffer[offset+44 : offset+48])
+	ctx.PidID = binary.LittleEndian.Uint32(decoder.buffer[offset+48 : offset+52])
+	_ = copy(ctx.Comm[:], decoder.buffer[offset+52:offset+68])
+	_ = copy(ctx.UtsName[:], decoder.buffer[offset+68:offset+84])
+	ctx.EventID = int32(binary.LittleEndian.Uint32(decoder.buffer[offset+84 : offset+88]))
+	ctx.Retval = int64(binary.LittleEndian.Uint64(decoder.buffer[offset+88 : offset+96]))
+	ctx.StackID = binary.LittleEndian.Uint32(decoder.buffer[offset+96 : offset+100])
+	ctx.Argnum = uint8(binary.LittleEndian.Uint16(decoder.buffer[offset+100 : offset+102]))
+	decoder.cursor += ctx.GetSizeBytes()
```

Comment:
Do we really need all of those GetSizeBytes() functions? To me it seems that they give us no real benefit as anyway the exact bytes offsets are hardcoded here. In addition, we will have to maintain those functions correct whenever we update one of the structs, increasing the chances of future programming mistakes.
If the problem it solves is padding (for example context is of size 104 while we read till byte 102 here), we can just add another field to the relevant structs of padding (for example, two more bytes for padding at the end of Context).

Reply:
Yep, I agree with @yanivagman. I just left a comment about the hard coded values asking if we could get the arg type's sizes during initialization (using reflect type size or something like that) and have them constant... even if we need to adjusting padding on the affected structs...

Reply:
@yanivagman , I'm fine with removing the `GetSizeBytes()` functions. I wouldn't add additional data to the structs just for expose the size of it. I would simply add it hard-coded.
@rafaeldtinoco I prefer not to use reflection as it negatively affects efficiency (`binary.Read` uses it, and that's the whole issue with it).

Reply:
we could still have GetSizeBytes() which returns an hardcoded value, and in the tests (https://github.com/aquasecurity/tracee/pull/1405/files#diff-242d9e577032010728b6349d850cb4bcb8c517923dd77d50ad684732b43c23efR31) we use reflection to get the size of the structs (`unsafe.Sizeof` doesn't help, I tried it already). If another developer adds/remove fields without updating the decoder then it will fail in tests.

Reply:
> @rafaeldtinoco I prefer not to use reflection as it negatively affects efficiency (`binary.Read` uses it, and that's the whole issue with it).

Yep, it was just a way to say something else than hardcoded in here. Not necessarily needs to be reflect.

Reply:
```
=== RUN   TestVfsWriteMetaSize
    protocol_test.go:31:
        	Error Trace:	protocol_test.go:31
        	Error:      	Not equal:
        	            	expected: 28
        	            	actual  : 20
        	Test:       	TestVfsWriteMetaSize
--- FAIL: TestVfsWriteMetaSize (0.00s)
=== RUN   TestKernelModuleMetaSize
--- PASS: TestKernelModuleMetaSize (0.00s)
=== RUN   TestMprotectWriteMetaSize
--- PASS: TestMprotectWriteMetaSize (0.00s)
FAIL
FAIL	github.com/aquasecurity/tracee/tracee-ebpf/tracee/internal/bufferdecoder	0.005s
```

After adding a variable to the arg type. I'm fine with that, since the PR checks will get the issue and allow the fix to happen (and I can understand why hard coding the size has better performance).